### PR TITLE
llvm-amdgpu: update development branch

### LINKED ISSUE #46986: llvm-amdgpu@master should point to 'amd-staging' branch instead of 'amd-stg-open'?
I think `spack install llvm-amdgpu@master` should point to the `amd-staging` branch, as that seems to be where AMD is adding new development (it is listed as the default branch): https://github.com/ROCm/llvm-project/branches

However, `package.py` points to the `amd-stg-open` branch:
https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/llvm-amdgpu/package.py#L27

However, this branch was last updated 10 months ago: https://github.com/ROCm/llvm-project/tree/amd-stg-open.

cc @srekolam @renjithravindrankannath @haampie

### PR DESCRIPTION
**Pull request by BenWibking**
This updates `llvm-amdgpu@master` to use `amd-staging`, since that's now the default branch for the GitHub repo: https://github.com/ROCm/llvm-project/branches.

Someone from AMD should confirm this is the right thing to do.

Closes https://github.com/spack/spack/issues/46986.

### CODE REVIEW

**Review by afzpatel — APPROVED** [review]
LGTM, amd-staging is the default branch. It seems that many of the ROCm recipes are pointing to deprecated branches and need to be corrected.

**Review by renjithravindrankannath — APPROVED** [review]

**Comment by afzpatel** [comment]
fyi, I created another pr to correct some of the other ROCm recipes: https://github.com/spack/spack/pull/47360

**Review by bernhardkaindl — DISMISSED** [review]

**Review by bernhardkaindl — COMMENTED** [review]
@BenWibking I ran a test build and got:

```py
  >> 276    CMake Error at .stage/spack-stage-llvm-amdgpu-master-jxj6okhnn7pf2wfx6lqybjn4q7357w35/spack-src/clang/CMakeLists.txt:211 (message):
     277      GCC_INSTALL_PREFIX is deprecated and will be removed.  Use configuration
     278      files (https://clang.llvm.org/docs/UsersManual.html#configuration-files)to
     279      specify the default --gcc-install-dir= or --gcc-triple=.  --gcc-toolchain=
     280      is discouraged.  See https://github.com/llvm/llvm-project/pull/77537 for
     281      detail.
```

It looks for building with `%gcc`, like you'll need to add a special case when spec satisfies `@master`, GCC_INSTALL_PREFIX is not set and `--gcc-install-dir=` or `--gcc-triple=` is used instead.

This probably refers to: https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/llvm-amdgpu/package.py#L258

**Comment by BenWibking** [comment]
I tried to do a test build, but I ran into a "failed to concretize" error: https://github.com/spack/spack/issues/47421.

**Review by bernhardkaindl — COMMENTED** [review]
@BenWibking Please see https://github.com/spack/spack/issues/47421#issuecomment-2455822105:
I cannot reproduce the error. This may be a case of a stale repository cache after the git merge of the develop branch.

You may also do a fresh clone of spack.git to a fresh directory and check the command in the new directory.

**Comment by BenWibking** [comment]
> @BenWibking I ran a test build and got:
> 
> ```python
>   >> 276    CMake Error at .stage/spack-stage-llvm-amdgpu-master-jxj6okhnn7pf2wfx6lqybjn4q7357w35/spack-src/clang/CMakeLists.txt:211 (message):
>      277      GCC_INSTALL_PREFIX is deprecated and will be removed.  Use configuration
>      278      files (https://clang.llvm.org/docs/UsersManual.html#configuration-files)to
>      279      specify the default --gcc-install-dir= or --gcc-triple=.  --gcc-toolchain=
>      280      is discouraged.  See https://github.com/llvm/llvm-project/pull/77537 for
>      281      detail.
> ```
> 
> It looks for building with `%gcc`, like you'll need to add a special case when spec satisfies `@master`, GCC_INSTALL_PREFIX is not set and `--gcc-install-dir=` or `--gcc-triple=` is used instead.
> 
> This probably refers to: https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/llvm-amdgpu/package.py#L258

This part of the package doesn't make any sense to me. `--gcc-install-dir` is actually a Clang command-line option, not a gcc one (very long discussion here: https://discourse.llvm.org/t/add-gcc-install-dir-deprecate-gcc-toolchain-and-remove-gcc-install-prefix/65091?page=2).

When using gcc to build llvm-amdgpu, it should do nothing. Since it did nothing to begin with, I've removed it from the package specification.

**Comment by BenWibking** [comment]
On Frontier, LLVM successfully builds by doing `spack install llvm-amdgpu@master`, but when I try to use it to compile my code, I get lots of errors like this when building device code:
```
/usr/lib64/gcc/x86_64-suse-linux/13/../../../../include/c++/13/array:203:9: error: reference to __host__ function 'operator[]' in __host__ __device__ function
  203 |         return _M_elems[__n];
      |                ^
```
which seems odd?

**Comment by BenWibking** [comment]
@bernhardkaindl Could you re-review?

**Review by bernhardkaindl — APPROVED** [review]
> @bernhardkaindl Could you re-review?

Hello @afzpatel, @renjithravindrankannath and @srekolam (maintainers), I generated some build results:<br>
:thumbsup::information_source: `spack install` on the changed recipes of this PR was successful[^1] on Ubuntu 22.04.4 LTS!<details><summary>Click here for a summary of the successful builds.</summary>
These changed specs were found using [gh pr diff](https://cli.github.com):
- `llvm-amdgpu@=master`


<details><summary><b>Click here to show/hide the report on </b><tt>spack find</tt> for <tt>linux-ubuntu22.04-skylake / gcc@13.1.0</tt>.</summary><ol>

<li><details><summary>Click on this area to unfold/refold the dependency report of this build:

>```py
>llvm-amdgpu@master~ipo~link_llvm_dylib~llvm_dylib+rocm-device-libs build_system=cmake build_type=Release generator=ninja patches=b4774ca
>```
</summary>

>```py
>	cmake@3.30.5
>		curl@8.10.1 tls=openssl
>			nghttp2@1.63.0
>		gmake@4.4.1
>	gcc-runtime@13.1.0
>	glibc@2.35
>	ncurses@6.5
>	ninja@1.12.1re2c
>		re2c@3.1
>	pkgconf@2.2.0
>	python@3.11.9bz2cryptctypesdbm,b0615b2,ebdca64,f2fd060
>		bzip2@1.0.8
>			diffutils@3.10
>		expat@2.6.3libbsd
>			libbsd@0.12.2
>				libmd@1.0.4
>		gdbm@1.23
>		gettext@0.22.5bzip2cursesgit
>			libxml2@2.10.3
>			tar@1.34 zip=pigz
>				pigz@2.8
>				zstd@1.5.6programs
>		libffi@3.4.6
>		libxcrypt@4.4.35_api
>			perl@5.40.0cpanmopcodeopenthreads
>				berkeley-db@18.1.40cxx,b231fcc
>		openssl@3.3.1 certs=mozilla
>			ca-certificates-mozilla@2023-05-30
>		readline@8.2
>		sqlite@3.46.0column_metadatadynamic_extensionsfts
>		util-linux-uuid@2.40.2
>		xz@5.4.6
>	z3@4.12.4
>		cmake@3.30.5
>			curl@8.10.1 tls=openssl
>				nghttp2@1.63.0
>			ncurses@6.5
>			zlib-ng@2.2.1compat_strategiesopt
>		gmake@4.4.1
>		python@3.10.14bz2cryptctypesdbm,7d40923,ebdca64,f2fd060
>			bzip2@1.0.8
>				diffutils@3.10
>			expat@2.6.3libbsd
>				libbsd@0.12.2
>					libmd@1.0.4
>			gdbm@1.23
>			gettext@0.22.5bzip2cursesgit
>				libxml2@2.10.3
>				tar@1.34 zip=pigz
>					pigz@2.8
>					zstd@1.5.6programs
>			libffi@3.4.6
>			libxcrypt@4.4.35_api
>				perl@5.40.0cpanmopcodeopenthreads
>					berkeley-db@18.1.40cxx,b231fcc
>			openssl@3.3.1 certs=mozilla
>				ca-certificates-mozilla@2023-05-30
>			pkgconf@2.2.0
>			readline@8.2
>			sqlite@3.46.0column_metadatadynamic_extensionsfts
>			util-linux-uuid@2.40.2
>			xz@5.4.6
>	zlib-ng@2.2.1compat_strategiesopt
>
>==> 1 installed package
>```

</details>



</details>

[^1]: Orchestrated using `spack install` by [github.com/spack/gh-spack-pr](https://github.com/spack/gh-spack-pr) and the [GitHub CLI](https://cli.github.com/)
</details><details><summary>Notes for @srekolam (maintainer)</summary>

This is not official, it is just based on what I know:
- If you like to review but need more time, please add a comment to the PR.
- If you review, you can add 👍, Approve, LGTM or Request in the comment.
- If you have questions (or to be removed as maintainer), please let me know.

You can also join the Spack Slack channel for more information and have a chat:
https://spackpm.slack.com → #pull-requests. Thank you and best regards, @bernhardkaindl</details>

**Review by bernhardkaindl — COMMENTED** [review]
@BenWibking:
> On Frontier, LLVM successfully builds by doing `spack install llvm-amdgpu@master`, but when I try to use it to compile my code, I get lots of errors like this when building device code:
> ```
> /usr/lib64/gcc/x86_64-suse-linux/13/../../../../include/c++/13/
> array:203:9: error: reference to __host__ function 'operator[]' in __host__ __device__ function
>    203 |         return _M_elems[__n];
>        |                ^
> ```
>  which seems odd?

No idea, maybe the compatibility with the GCC-13 headers of the host is not great.
However, that's not a spack build issue. I'm not the right person to investigate that.

**Review by bernhardkaindl — COMMENTED** [review]
Regarding the no longer supported GCC_INSTALL_PREFIX in the development branch:

#43795 is related. Maybe we could still need GCC_INSTALL_PREFIX to build the released versions of `llvm-amdgpu`. That needs to be build-tested.

And @afzpatel and @renjithravindrankannath's approvals were before this change, so formally, one of them should also re-review or test and approve.

Unless the change to the GCC_INSTALL_PREFIX is restricted to changing the development version only, I'd assign this to `haampie` for final review and merge.

**Comment by BenWibking** [comment]
> Unless the change to the GCC_INSTALL_PREFIX is restricted to changing the development version only, I'd assign this to `haampie` for final review and merge.

The package sets a Clang-specific option when gcc is used to build. I don't see how this is doing the right thing for *any* version, so yes, I'd like someone to take a close look at it.

**Review by afzpatel — APPROVED** [review]
LGTM, I think @haampie should review and approve before merging this. If it's true that --gcc-install-dir doesn't do anything when building llvm-amdgpu with gcc, then it should be removed from the llvm recipe as well.

**Comment by haampie** [comment]
The TL;DR for GCC_INSTALL_PREFIX:

It's deprecated, but had the upside that it would fix libstdc++ also when building runtimes (which use just-built clang as a compiler). The downside was that you could not be specific about what installation to use from /usr if you had multiple GCCs there.

Now they added a flag so you can select a specific GCC when multiple are in a single prefix, but removed the option to bake that path into the binary. So, we cannot guarantee things work well during the builds of runtimes, and we need to generate config files instead.

llvm-amdgpu should follow llvm's implementation of that.

**Comment by alecbcs** [comment]
@BenWibking is this PR still relevant? I see there's a conflict and the package has since been updated on develop.

**Comment by BenWibking** [comment]
> @BenWibking is this PR still relevant? I see there's a conflict and the package has since been updated on develop.

I don't know, but I don't have a need to build llvm-amdgpu from scratch anymore, so I'll close this.
